internal/pkg/vm/internal/vmcontext/invocation_context.go

```diff
@@ -0,0 +1,308 @@
+	if ctx.msg.method == types.SendMethodID {
+		return message.Ok().WithGas(ctx.gasTank.GasConsumed())
+	}
+
+	// 4. load target actor
+	// Note: we replace the "to" address with the normalized version
+	ctx.toActor, ctx.msg.to = ctx.resolveTarget(ctx.msg.to)
+
+	// 5. load target actor code
+	// TODO: use chain height based protocol version here (#3360)
+	actorImpl := ctx.rt.getActorImpl(ctx.toActor.Code)
+
+	// 6. create target state handle
+	stateHandle := newActorStateHandle((*stateHandleContext)(ctx), ctx.toActor.Head)
+	ctx.stateHandle = &stateHandle
+
+	// dispatch
+	// 2. check method exists
+	// 3. invoke method on actor
+
+	// 2. check method exists
+	exportedFn, ok := makeTypedExport(actorImpl, ctx.msg.method)
+	if !ok {
+		panic(exitcode.InvalidMethod)
+	}
+
+	// 3. invoke method on actor
+	vals, code, err := exportedFn(ctx)
+
+	// Handle legacy errors and codes
+	if err != nil {
+		runtime.Abort("Legacy actor code returned an error")
+	}
+	if code != 0 {
+		runtime.Abort("Legacy actor code returned with non-zero error code")
+	}
+
+	// post-dispatch
+	// 1. check caller was validated
+	// 2. check state manipulation was valid
+	// 3. success!
+
+	// 1. check caller was validated
+	if !ctx.isCallerValidated {
+		runtime.Abort("Caller MUST be validated during method execution")
+	}
+
+	// 2. validate state access
+	ctx.stateHandle.Validate()
+
+	// 3. success! build the receipt
+	if len(vals) > 0 {
+		return vals[0]
+	}
+	return nil
+}
+
+// resolveTarget loads and actor and returns its ActorID address.
+//
+// If the target actor does not exist, and the target address is a pub-key address,
+// a new account actor will be created.
+// Otherwise, this method will abort execution.
+func (ctx *invocationContext) resolveTarget(target address.Address) (*actor.Actor, address.Address) {
+	// resolve the target address via the InitActor, and attempt to load state.
+	initActorEntry, err := ctx.rt.state.GetActor(context.Background(), address.InitAddress)
+	if err != nil {
+		panic("init actor not found")
+	}
+
+	// build state handle
+	var stateHandle = NewReadonlyStateHandle(ctx.rt.LegacyStorage(), initActorEntry.Head)
+
+	// get a view into the actor state
+	initView := initactor.NewView(stateHandle, ctx.rt.LegacyStorage())
+
+	// lookup the ActorID based on the address
+	targetIDAddr, ok := initView.GetIDAddressByAddress(target)
+	if ok {
+		targetActor, err := ctx.rt.state.GetActor(context.Background(), targetIDAddr)
+		if err == nil {
+			// actor found, return it and its IDAddress
+			return targetActor, targetIDAddr
+		}
+	}
+
+	// actor does not exist, create an account actor
+	// - precond: address must be a pub-key
+	// - sent init actor a msg to create the new account
+
+	if !target.IsPubKey() {
+		// Don't implicitly create an account actor for an address without an associated key.
+		panic(exitcode.ActorNotFound)
+	}
+
+	// send init actor msg to create the account actor
```

Comment:
This differs from the spec at present. Leave a comment pointing to https://github.com/filecoin-project/specs/issues/762

Reply:
Quite a few things changed since mid December, I'll do another pass on a separate PR to bring it up to date again.